server/enterprise/message_export/shared/export_data.go

```diff
@@ -0,0 +1,439 @@
+		if err != nil {
+			return nil, results, err
+		}
+		uploadStartsByChannel[channelId] = append(uploadStartsByChannel[channelId], startUploads...)
+		uploadStopsByChannel[channelId] = append(uploadStopsByChannel[channelId], stopUploads...)
+		deletedFilesByChannel[channelId] = append(deletedFilesByChannel[channelId], deleteFileMessages...)
+		filesByChannel[channelId] = append(filesByChannel[channelId], uploadedFiles...)
+
+		if err := metadata.UpdateCounts(channelId, 1, len(uploadedFiles)); err != nil {
+			return nil, results, err
+		}
+
+		if _, ok := postAuthorsByChannel[channelId]; !ok {
+			postAuthorsByChannel[channelId] = make(map[string]ChannelMember)
+		}
+		postAuthorsByChannel[channelId][*post.UserId] = ChannelMember{
+			UserId:   *post.UserId,
+			Email:    *post.UserEmail,
+			Username: *post.Username,
+			IsBot:    post.IsBot,
+		}
+	}
+
+	// If the channel is not in channelsInThisBatch (i.e. if it didn't have a post), we need to check if it had
+	// user activity between this batch's startTime-endTime. If so, add it to the channelsInThisBatch.
+	for id := range p.ChannelMetadata {
+		if !channelsInThisBatch[id] {
+			if ChannelHasActivity(p.ChannelMemberHistories[id], p.BatchStartTime, p.BatchEndTime) {
+				channelsInThisBatch[id] = true
+			}
+		}
+	}
+
+	// Build the channel exports for the channels that had post or user join/leave activity this batch.
+	channelExports := make([]ChannelExport, 0, len(channelsInThisBatch))
+	for id := range channelsInThisBatch {
+		c := metadata.Channels[id]
+
+		joinEvents, leaveEvents := getJoinsAndLeaves(p.BatchStartTime, p.BatchEndTime,
+			p.ChannelMemberHistories[id], postAuthorsByChannel[id])
+
+		channelExports = append(channelExports, ChannelExport{
+			ChannelId:    c.ChannelId,
+			ChannelType:  c.ChannelType,
+			ChannelName:  c.ChannelName,
+			DisplayName:  c.ChannelDisplayName,
+			StartTime:    p.BatchStartTime,
+			EndTime:      p.BatchEndTime,
+			Posts:        postsByChannel[id],
+			Files:        filesByChannel[id],
+			DeletedFiles: deletedFilesByChannel[id],
+			UploadStarts: uploadStartsByChannel[id],
+			UploadStops:  uploadStopsByChannel[id],
+			JoinEvents:   joinEvents,
+			LeaveEvents:  leaveEvents,
+		})
+	}
+
+	return channelExports, results, nil
+}
+
+// postToAttachmentsEntries returns every fileInfo as uploadedFiles. It also adds each file into the lists:
+//
+//	startUploads, stopUploads, and deleteFileMessages (for ActianceExport).
+func postToAttachmentsEntries(post *model.MessageExport, db MessageExportStore) (
+	uploadedFiles []*model.FileInfo, startUploads []*FileUploadStartExport, stopUploads []*FileUploadStopExport, deleteFileMessages []PostExport, err error) {
+	// if the post included any files, we need to add special elements to the export.
+	if len(post.PostFileIds) == 0 {
+		return
+	}
+
+	uploadedFiles, err = db.FileInfo().GetForPost(*post.PostId, true, true, false)
+	if err != nil {
+		return
+	}
```

Comment:
Old logic, but yah I think there's a better way to handle these or at least surface them. I haven't gone through all the ways this can fail yet, that's for the "handle errors better" https://mattermost.atlassian.net/browse/MM-61993 and reliability/restart pass https://mattermost.atlassian.net/browse/MM-59093